src/enzo/ActiveParticleRoutines.C

```diff
@@ -0,0 +1,508 @@
+  GridID = a->GridID;
+  type = a->type;
+  return;
+}
+
+/**********************
+
+   CONVENIENT ROUTINES
+
+ **********************/
+
+template<class active_particle_class>
+active_particle_class *ActiveParticleType::copy(void)
+{
+  int i, dim;
+  active_particle_class *a = new active_particle_class();
+  a->CurrentGrid = CurrentGrid;
+  for (dim = 0; dim < MAX_DIMENSION; dim++) {
+    a->pos[dim] = pos[dim];
+    a->vel[dim] = vel[dim];
+  }
+  a->Mass = Mass;
+  a->BirthTime = BirthTime;
+  a->DynamicalTime = DynamicalTime;
+  a->Metallicity = Metallicity;
+  a->Identifier = Identifier;
+  a->level = level;
+  a->GridID = GridID;
+  a->type = type;
+  return a;
+}
+
+void ActiveParticleType::ConvertAllMassesToSolar(void)
+{
+  const double Msun = 1.989e33;
+  double dx;
+  float DensityUnits, LengthUnits, TemperatureUnits, TimeUnits,
+    VelocityUnits, MassConversion;
+  GetUnits(&DensityUnits, &LengthUnits, &TemperatureUnits,
+	   &TimeUnits, &VelocityUnits, CurrentGrid->Time);
+  dx = LengthUnits * CurrentGrid->CellWidth[0][0];
+  MassConversion = (float) (dx*dx*dx * double(DensityUnits) / Msun);
+  this->Mass *= MassConversion;
+  return;
+}
+
+void ActiveParticleType::ConvertMassToSolar(void)
+{
+  const double Msun = 1.989e33;
+  double dx;
+  float DensityUnits, LengthUnits, TemperatureUnits, TimeUnits,
+    VelocityUnits, MassConversion;
+  GetUnits(&DensityUnits, &LengthUnits, &TemperatureUnits,
+	   &TimeUnits, &VelocityUnits, CurrentGrid->Time);
+  dx = LengthUnits * CurrentGrid->CellWidth[0][0];
+  MassConversion = (float) (dx*dx*dx * double(DensityUnits) / Msun);
+  this->Mass *= MassConversion;
+  return;
+}
+
+void  ActiveParticleType::AdjustVelocity(float VelocityIncrement[])
+{ 
+  int i;
+  for (i = 0; i<3; i++)
+    vel[i] += VelocityIncrement[i];
+  return;
+}
+
+void ActiveParticleType::SetVelocity(float NewVelocity[])
+{
+  int i;
+  for (i = 0; i<3; i++)
+    vel[i] = NewVelocity[i];
+  return;
+}
+
+void ActiveParticleType::SetPosition(FLOAT NewPosition[])
+{
+  int i;
+  for (i = 0; i<3; i++)
+    pos[i] = NewPosition[i];
+  return;
+}
+
+void ActiveParticleType::SetPositionPeriod(FLOAT period[])
+{
+  int i;
+  for (i = 0; i<3; i++) {
+    pos[i] = fmod(pos[i], period[i]);
+    if (pos[i] < 0) {
+      pos[i] += period[i];
+    }
+  }
+  return;
+}
+
+void ActiveParticleType::Merge(ActiveParticleType *a)
+{
+  int dim;
+  double ratio1, ratio2;
+  ratio1 = Mass / (Mass + a->Mass);
+  ratio2 = 1.0 - ratio1;
+  Metallicity = ratio1 * Metallicity + ratio2 * a->Metallicity;
+  for (dim = 0; dim < MAX_DIMENSION; dim++) {
+    pos[dim] = ratio1 * pos[dim] + ratio2 * a->pos[dim];
+    vel[dim] = ratio1 * vel[dim] + ratio2 * a->vel[dim];
+    //accreted_angmom[dim] = ratio1 * accreted_angmom[dim] + ratio2 * a.accreted_angmom[dim];
+  }
+  Mass += a->Mass;
+  //FinalMass += a.FinalMass;
+  //DeltaMass += a.DeltaMass;
+  //last_accretion_rate += a.last_accretion_rate;
+  //NotEjectedMass += a.NotEjectedMass;
+  return;
+}
```

Comment:
Many commented out lines in this function. Slight clean will make it look better.

Reply:
Done